### PR TITLE
[UAP] Roles and collections

### PR DESCRIPTION
**Pull request by MCGallaspy**
## Summary

Continue implementing the UAP backend. Looking back, I'm not finding the commit names super informative -- happy to amend the messages for this PR at request.

## TODO

- [x] Have tests been written for the new code?
- [x] Has documentation been written/updated? **Updates existing docs page through automodule directive**
- [x] New dependencies (if any) added to requirements file
- [x] Add an entry to CHANGELOG.rst

## Reviewer guidance

4 new Models added, which is the main action of this PR. Their purpose is described in the technical spec (see documentation section below). This PR mainly implements `HierarchyNode`, with following PRs implementing `Role` and `Collection`, as well as permissions checking.

## Documentation

See the [UAP tech spec](https://docs.google.com/document/d/1BcOSxhoAOZGKjg9e7C6PpeQfHj9zkDqrKIk1WrbLsDU/edit#heading=h.fzctnhrv2u2t) for more info.

### CODE REVIEW

**Comment by MCGallaspy** [comment]
Strange, [this error](https://travis-ci.org/learningequality/kolibri/jobs/106631961) was fixed for me (using python2.7) by rebuilding the tox virtualenv.

**Comment by MCGallaspy** [comment]
@benjaoming any clues as to the test failures?

**Comment by MCGallaspy** [comment]
TIL to exclude the `contenttypes` app when using Django's dumpdata command

File: .travis.yml [hunk 1]
```diff
@@ -26,7 +26,7 @@ before_script:
 
 # command to run tests, e.g. python setup.py test
 script:
-  - tox -e $TOX_ENV
+  - tox -e $TOX_ENV -r
```
Comment (benjaoming) [inline .travis.yml:29]: Was this necessary? I initially wanted tox to reuse the first environment to generate our docs and run BDD.. not sure how well that played in the end.

**Comment by benjaoming** [comment]
@MCGallaspy regarding test fixtures: They need maintenance. It might also be difficult to sense the test data because of the JSON format. I agree that tests creating lots of object instances are hard to read. On the other hand, they can be generated in a more dynamic fashion.

What's your take on this? How would you prefer testing to look?

**Comment by MCGallaspy** [comment]
> What's your take on this? How would you prefer testing data to be managed in the longer run?

Ideally I'd like to generate a fixture once and never have to worry about it again (unless the schema of the specific models I'm testing changes). I considered writing a management command to generate fixtures on demand, but wondered if it might be considered bloat?

Though perhaps I'm wrong and fixtures *must* be periodically updated, as you're suggesting? In which case a management command to generate/update all fixtures might not be a bad idea.

File: .travis.yml (hunk 1 above)
Comment (MCGallaspy) [inline .travis.yml:29]: Not really sure, I'll try removing it.

Comment (MCGallaspy) [inline .travis.yml:29]: Though at least in one instance locally, tox failed to install new requirements when I added them to the req's file. So probably it is necessary.

Comment (benjaoming) [inline .travis.yml:29]: Not for travis though? It doesn't cache anything. The intention was to make travis speedier by having it reuse stuff :)

**Comment by MCGallaspy** [comment]
Don't merge. Re-opening using a different implementation.

**Pull request closed without merging**